[docs] troubleshooting docs update (#2326)
open-telemetry/opentelemetry-dotnet-instrumentation: docs/troubleshooting.md

# Troubleshooting

## General steps

> If you encounter any issue with OpenTelemetry .NET Automatic Instrumentation,

Reviewer: ```suggestion
If you encounter any issue with OpenTelemetry .NET automatic instrumentation,
```

there are steps that can help you understand the issue.

### Enable detailed logging

Detailed debug logs can help you troubleshoot instrumentation issues, and can be
attached to issues in this project to facilitate investigation.

To get the detailed logs from the OpenTelemetry .NET Automatic Instrumentation, set
the [`OTEL_LOG_LEVEL`](./config.md#internal-logs) environment variable to `debug`
before the instrumented process starts.

By default, the library writes the log files under predefined [locations](./config.md#internal-logs).
If needed, change the default location by updating the `OTEL_DOTNET_AUTO_LOG_DIRECTORY`
environment variable.

After obtaining the logs, remove the `OTEL_LOG_LEVEL`
environment variable, or set it to less verbose level
to avoid unnecessary overhead.

### Enable host tracing

[Host tracing](https://github.com/dotnet/runtime/blob/edd23fcb1b350cb1a53fa409200da55e9c33e99e/docs/design/features/host-tracing.md#host-tracing)
can be used to gather the information needed to investigate the problems

> related to e.g assemblies not being found. Set the following environment variables:

Reviewer: ```suggestion
related to various issues, like assemblies not being found. Set the following environment
variables:
```

```terminal
COREHOST_TRACE=1
COREHOST_TRACEFILE=corehost_verbose_tracing.log
```

> and re-run the application to collect the log.

Reviewer: ```suggestion
Then restart the application to collect the logs.
```

## Common issues

> ### Nothing happens

Reviewer: Can we be more specific? Like "No telemetry is produced", for example?

#### Symptoms

There is no telemetry generated.

> There are no logs in OpenTelemetry .NET Automatic Instrumentation internal logs [location](./config.md#internal-logs).

Reviewer: ```suggestion
There are no logs in OpenTelemetry .NET automatic instrumentation internal logs [location](./config.md#internal-logs).
```

No strong feelings about this, but I wouldn't capitalize "automatic instrumentation".

Author: Capitalized version is used extensively in docs, I prefer to keep it as it is for consistency.

> It may occur that the .NET Profiler is unable to attach

Reviewer: ```suggestion
It might occur that the .NET Profiler is unable to attach
```

Reviewer: Attach to...?

and therefore no logs would be emitted.

#### Solution

The most common reason is that the instrumented application
has no permissions to load the OpenTelemetry .NET Automatic Instrumentation

> assemblies.

Reviewer: How would the user provide permissions?

### Performance issues

#### Symptoms

> High CPU usage.

Reviewer: ...when running the instrumented service? How do we define high? I know it's tricky, but we might want to provide an example based on some existing issue.

Reviewer: Also, any way they can reliably check CPU usage from PowerShell?

#### Solution

Make sure that you have not enabled the automatic instrumentation globally
by setting the environment variables at system or user scope.

> If the system or user scope is intended, use the [`OTEL_DOTNET_AUTO_EXCLUDE_PROCESSES`](./config.md#global-settings)

Reviewer: ```suggestion
If the usage of system or user scope is intentional, use the [`OTEL_DOTNET_AUTO_EXCLUDE_PROCESSES`](./config.md#global-settings)
```

environment variables to exclude applications from the automatic instrumentation.

### `dotnet` CLI tool is crashing

#### Symptoms

> Error message when running an app with e.g `dotnet run` similar to the one below:

Reviewer: ```suggestion
You get error messages similar to the one below when running an app, for example with `dotnet run`:
```

```txt
PS C:\Users\Administrator\Desktop\OTelConsole-NET6.0> dotnet run My.Simple.Console
Unhandled exception. System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
---> System.TypeInitializationException: The type initializer for 'OpenTelemetry.AutoInstrumentation.Loader.Startup' threw an exception.
---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
---> System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.Extensions.Configuration.Abstractions, Version=7.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.
```

#### Related issues

> - [#1744](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/1744)

Reviewer: This is great. Could we add related issues for all other issues as well?

#### Solution

> Until version `v0.6.0-beta.1` (inclusive) there were issues instrumenting

Reviewer: ```suggestion
With version `v0.6.0-beta.1` and lower, there were issues when instrumenting
```

Reviewer: Some sample issues in the tracker?

# Troubleshooting

## General steps

> If you encounter any issue with OpenTelemetry .NET automatic instrumentation,

Reviewer 2: it is against capitalization convention ; )

Author: fixed in https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/2326/commits/5ab2e8fbdeced545c73045d37380285a09da95e0

Unhandled exception. System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.Extensions.DependencyInjection.Abstractions, Version=7.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.

File name: 'Microsoft.Extensions.DependencyInjection.Abstractions, Version=7.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'
   at Microsoft.AspNetCore.Builder.WebApplicationBuilder..ctor(WebApplicationOptions options, Action`1 configureDefaults)
   at Microsoft.AspNetCore.Builder.WebApplication.CreateBuilder(String[] args)
   at Program.<Main>$(String[] args) in /Blog.Core/Blog.Core.Api/Program.cs:line 26
```

#### Related issues

- [#2269](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2269)
- [#2296](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2296)

#### Solution

OpenTelemetry .NET NuGet packages and its dependencies
are deployed with the OpenTelemetry .NET Automatic Instrumentation.

To handle dependency versions conflicts,
update the instrumented application's project references
to use the same versions as OpenTelemetry .NET Automatic Instrumentation.

> Dependencies used by OpenTelemetry .NET Automatic Instrumentation can be found at:

Reviewer: ```suggestion
The following dependencies are used by OpenTelemetry .NET automatic instrumentation:
```

- [OpenTelemetry.AutoInstrumentation](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/src/OpenTelemetry.AutoInstrumentation/OpenTelemetry.AutoInstrumentation.csproj)
- [OpenTelemetry.AutoInstrumentation.AdditionalDeps](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/src/OpenTelemetry.AutoInstrumentation.AdditionalDeps/Directory.Build.props)

> Their versions can be found at:

Reviewer: ```suggestion
Find their versions in the following locations:
```

- [Directory.Packages.props](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/Directory.Packages.props)
- [src/Directory.Packages.props](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/src/Directory.Packages.props)
- [src/OpenTelemetry.AutoInstrumentation.AdditionalDeps/Directory.Packages.props](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/src/OpenTelemetry.AutoInstrumentation.AdditionalDeps/Directory.Packages.props)

> For .NET Framework applications the assembly references are, by default, updated
> during runtime to the versions used by the automatic instrumentation.

Reviewer: ```suggestion
By default, assembly references for .NET Framework applications are upgraded
during runtime to the versions used by the automatic instrumentation.
```

> This behavior can be controlled via the [`OTEL_DOTNET_AUTO_NETFX_REDIRECT_ENABLED`](./config.md#additional-settings)

Reviewer: ```suggestion
This behavior can be controlled through the [`OTEL_DOTNET_AUTO_NETFX_REDIRECT_ENABLED`](./config.md#additional-settings)
```

For more information about the GAC usage by the automatic instrumentation,
see [here](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/1906#issuecomment-1376292814).

### Assembly in AdditionalDeps was not found

#### Symptoms

> Error message similar to the one below:

Reviewer: ```suggestion
You get an error message similar to the following:
```

```txt
An assembly specified in the application dependencies manifest (OpenTelemetry.AutoInstrumentation.AdditionalDeps.deps.json) was not found  
```

#### Related issues

- [#1744](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/1744)
- [#2181](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/2181)

#### Solution

> If issue you encountered is not one of the issues above, see general steps
> at the top to collect additional info to help facilitate issue investigation.

Reviewer: ```suggestion
If you encounter an issue not listed on this page, see [General steps](#general-steps)
to collect additional diagnostic information. This might help facilitate troubleshooting.
```

Would it be useful to link to issue guidelines?
